File: tests/e2e/negate.c
#include <stdio.h>

extern void negate_float(float *out, float *a);
extern void negate_int(int *out, int *a);

int main() {
    float fa = 42.0f, fout = 0.0f;
    negate_float(&fout, &fa);
    printf("%.1f\n", fout);

    int ia = 7, iout = 0;
    negate_int(&iout, &ia);
    printf("%d\n", iout);

    return 0;
}

File: tests/e2e/negate.py
# RUN: uv run xdsl-exo -o %t.mlir %s
# RUN: mlir-opt --convert-arith-to-llvm --convert-cf-to-llvm --convert-func-to-llvm --reconcile-unrealized-casts %t.mlir | mlir-translate --mlir-to-llvmir | llc -filetype=obj -o %t.o
# RUN: clang %t.o %S/negate.c -o %t.bin -lm
# RUN: %t.bin | filecheck %s

from __future__ import annotations

from exo import *


# CHECK: -42.0
# CHECK: -7
@proc
def negate_float(out: f32[1] @ DRAM, a: f32[1] @ DRAM):
    out[0] = -a[0]


@proc
def negate_int(out: i32[1] @ DRAM, a: i32[1] @ DRAM):
    out[0] = -a[0]
